lib/statsd.rb

```diff
@@ -14,14 +14,77 @@
 #   statsd = Statsd.new('localhost').tap{|sd| sd.namespace = 'account'}
 #   statsd.increment 'activate'
 class Statsd
-  class RubyUdpClient
-    attr_reader :key, :sock
+  class BufferedUDPClient < UDPClient
+    @max_packet_size = 511
+    (class << self; attr_accessor :max_packet_size; end)
 
-    def initialize(address, port, key = nil)
+    def initialize(address, port)
+      super(address, port)
+      @buffer = String.new
+    end
+
+    def flush!
+      return unless @buffer.bytesize > 0
+      sock.write(@buffer)
+      @buffer.clear
+    rescue => boom
+      nil
+    end
+
+    def send(msg)
+      flush! if @buffer.bytesize + msg.bytesize > BufferedUDPClient.max_packet_size
+      @buffer << msg
+      @buffer << "\n".freeze
+      nil
```

Comment:
> If `@buffer.bytesize + msg.bytesize == BufferedUDPClient.max_packet_size`, we won't `flush!` on this `send`. But because of the newline, the next `send` call will actually send `BufferedUDPClient.max_packet_size + 1` bytes.

Oh, I see now that this is intended because `BufferedUDPClient.max_packet_size` is 511, not 512. That seems a bit misleading; the maximum packet we want to send is 512 bytes, not 511, so it's weird for the constant to be 511.